src/nvim/event/pty_process.c

```diff
@@ -255,3 +202,50 @@ static void init_termios(struct termios *termios) FUNC_ATTR_NONNULL_ALL
   termios->c_cc[VMIN]     = 1;
   termios->c_cc[VTIME]    = 0;
 }
+
+static bool set_duplicating_descriptor(int fd, uv_pipe_t *pipe)
+  FUNC_ATTR_NONNULL_ALL
+{
+  int fd_dup = dup(fd);
+  if (fd_dup < 0) {
+    ELOG("Failed to dup descriptor %d: %s", fd, strerror(errno));
+    return false;
+  }
+  int uv_result = uv_pipe_open(pipe, fd_dup);
+  if (uv_result) {
+    ELOG("Failed to set pipe to descriptor %d: %s",
+         fd_dup, uv_strerror(uv_result));
+    close(fd_dup);
+    return false;
+  }
+  return true;
+}
+
+static void chld_handler(uv_signal_t *handle, int signum)
```

Comment:
They are handled in the same loop, but it's only possible to have `waitpid` return a pid > 0 once(at least in my tests). So if libuv `SIGCHLD` handler receives it first, this one will short circuit because the process will have already exited. The same will happen for libuv if we catch it first: https://github.com/libuv/libuv/blob/v1.x/src/unix/process.c#L78-L80